# Uvod v Pandas

## O Pandas-u


**Dimension & Description**



<table class="table table-bordered">
<tbody><tr>
<th style="text-align:center;">Data Structure</th>
<th style="text-align:center;">Dimensions</th>
<th style="text-align:center;">Description</th>
</tr>
<tr>
<td style="text-align:center;">Series</td>
<td style="text-align:center;">1</td>
<td style="text-align:center;">1D labeled homogeneous array, sizeimmutable.</td>
</tr>
<tr>
<td style="text-align:center;">Data Frames</td>
<td style="text-align:center;">2</td>
<td style="text-align:center;">General 2D labeled, size-mutable tabular structure with potentially heterogeneously typed
columns.</td>
</tr>
<tr>
<td style="text-align:center;">Panel</td>
<td style="text-align:center;">3</td>
<td style="text-align:center;">General 3D labeled, size-mutable array.</td>
</tr>
</tbody></table>


## Importing pandas

In [3]:
import pandas as pd
import numpy as np

Check the version:

In [2]:
pd.__version__

'0.24.2'

## Reminder about Built-In Documentation



For example, to display all the contents of the pandas namespace, you can type

In [ ]:
pd.<TAB>

And to display Pandas's built-in documentation, you can use this:

In [3]:
pd?

More detailed documentation, along with tutorials and other resources, can be found at http://pandas.pydata.org/.

## Introducing Pandas Objects



In [4]:
import numpy as np
import pandas as pd

### The Pandas Series Object


In [6]:
data = pd.Series([0.25,0.5,0.75,1.0])

In [7]:
data

0    0.25
1    0.50
2    0.75
3    1.00
dtype: float64

In [8]:
data[1]

0.5

#### Series as generalized NumPy array


#### Series as specialized dictionary

In [55]:
population_dict = {'California' : 383352256,
            'Texas' : 245578987,
            'New York' : 789635412,
            'Florida' : 123365478,
            'Illonois' : 658896234}
population = pd.Series(population_dict)
#print(population)
#print()
area_dict = {'California' : 123456789,
            'Texas' : 321654987,
            'New York' : 564897231,
            'Florida' : 456789123,
            'Illonois' : 963852741}
area = pd.Series(area_dict)
#print(area)
#print()

states = pd.DataFrame({'population' : population,'area' : area})
#states
#type(states)
#states.index
#states.values
#states.columns
pd.DataFrame(population, columns = ['columns'])
pd.DataFrame(np.random.rand(3,2),
            columns = ['foo', 'bar'],
            index = ['a','b','c'])
pd.DataFrame([{'a':1, 'b':2}, {'b':3, 'c':4}])



a  b    c
0  1.0  2  NaN
1  NaN  3  4.0

In [63]:
ind = pd.index=[2,3,5,7,12]
ind[1]

3

#### Constructing Series objects

`>>> pd.Series(data, index=index)`


[Dokumentacija](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.Series.html)

### The Pandas DataFrame Object



#### DataFrame as specialized dictionary


#### Constructing DataFrame objects



- **From a single Series object**



 - **From a dictionary of Series objects**
 - **From a two-dimensional NumPy array**
 


- **From a list of dicts**



[Dokumentacija](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.html)

### The Pandas Index Object



#### Index as immutable array



## Importing Data with Pandas

In [8]:
import numpy as np
import pandas as pd

!head -n 2 data/DATA_fortune_500_2017.csv

In [9]:
f500 = pd.read_csv('data/DATA_fortune_500_2017.csv', index_col = 0)

In [10]:
type(f500)

pandas.core.frame.DataFrame

In [12]:
f500.shape


(500, 22)

##  Introducing DataFrames



<p><img alt="anatomy of a dataframe" src="https://s3.amazonaws.com/dq-content/291/df_anatomy.svg"></p>



In [14]:
f500.dtypes

Title           object
Website         object
Employees        int64
Sector          object
Industry        object
Hqlocation      object
Hqaddr          object
Hqcity          object
Hqstate         object
Hqzip            int64
Hqtel           object
Ceo             object
Ceo-title       object
Address         object
Ticker          object
Fullname        object
Revenues         int64
Revchange      float64
Profits        float64
Prftchange      object
Assets           int64
Totshequity    float64
dtype: object

In [16]:
f500.head(2)

Title                           Website  Employees  \
Rank                                                                    
1                Walmart            http://www.walmart.com    2300000   
2     Berkshire Hathaway  http://www.berkshirehathaway.com     367700   

          Sector                                  Industry       Hqlocation  \
Rank                                                                          
1      Retailing                     General Merchandisers  Bentonville, AR   
2     Financials  Insurance: Property and Casualty (Stock)        Omaha, NE   

                   Hqaddr       Hqcity Hqstate  Hqzip  ...  \
Rank                                                   ...   
1     702 S.W. Eighth St.  Bentonville      AR  72716  ...   
2         3555 Farnam St.        Omaha      NE  68131  ...   

                                          Ceo-title  \
Rank                                                  
1     President, Chief Executive Officer & Director   
2                Chairman & Chief Executive Officer   

                                         Address Ticker  \
Rank                                                      
1     702 S.W. Eighth St., Bentonville, AR 72716    WMT   
2               3555 Farnam St., Omaha, NE 68131   BRKA   

                     Fullname Revenues Revchange  Profits  Prftchange  Assets  \
Rank                                                                            
1       Wal-Mart Stores, Inc.   485873       0.8  13643.0        -7.2  198825   
2     Berkshire Hathaway Inc.   223604       6.1  24074.0           0  620854   

     Totshequity  
Rank              
1        77798.0  
2       283001.0  

[2 rows x 22 columns]

In [18]:
f500.tail(5)

Title                      Website  Employees  \
Rank                                                           
496    Michaels Cos.      http://www.michaels.com      31000   
497    Toll Brothers  http://www.tollbrothers.com       4200   
498            Yahoo         http://www.yahoo.com       8500   
499    Vistra Energy  http://www.vistraenergy.com       4431   
500   ABM Industries           http://www.abm.com     110000   

                          Sector                          Industry  \
Rank                                                                 
496                    Retailing        Specialty Retailers: Other   
497   Engineering & Construction                      Homebuilders   
498                   Technology   Internet Services and Retailing   
499                       Energy                            Energy   
500            Business Services  Diversified Outsourcing Services   

         Hqlocation                Hqaddr     Hqcity Hqstate  Hqzip  ...  \
Rank                                                                 ...   
496      Irving, TX  8000 Bent Branch Dr.     Irving      TX  75063  ...   
497     Horsham, PA     250 Gibraltar Rd.    Horsham      PA  19044  ...   
498   Sunnyvale, CA        701 First Ave.  Sunnyvale      CA  94089  ...   
499      Dallas, TX        1601 Bryan St.     Dallas      TX  75201  ...   
500    New York, NY       1 Liberty Plaza   New York      NY  10006  ...   

                                          Ceo-title  \
Rank                                                  
496              Chairman & Chief Executive Officer   
497              Chief Executive Officer & Director   
498   President, Chief Executive Officer & Director   
499   President, Chief Executive Officer & Director   
500   President, Chief Executive Officer & Director   

                                     Address Ticker  \
Rank                                                  
496   8000 Bent Branch Dr., Irving, TX 75063    MIK   
497     250 Gibraltar Rd., Horsham, PA 19044    TOL   
498      701 First Ave., Sunnyvale, CA 94089   YHOO   
499         1601 Bryan St., Dallas, TX 75201    VST   
500      1 Liberty Plaza, New York, NY 10006    ABM   

                          Fullname Revenues Revchange  Profits  Prftchange  \
Rank                                                                         
496   The Michaels Companies, Inc.     5197       5.8    378.2         4.2   
497            Toll Brothers, Inc.     5170      23.9    382.1         5.2   
498                    Yahoo! Inc.     5169       4.0   -214.3         5.2   
499            Vistra Energy Corp.     5164       4.0   -214.3         5.2   
500    ABM Industries Incorporated     5145      -2.8     57.2         -25   

      Assets Totshequity  
Rank                      
496     2148     -1698.0  
497     9737      4229.0  
498    48083     31049.0  
499    15167      6597.0  
500     2281       974.0  

[5 rows x 22 columns]

In [19]:
f500.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 500 entries, 1 to 500
Data columns (total 22 columns):
Title          500 non-null object
Website        500 non-null object
Employees      500 non-null int64
Sector         500 non-null object
Industry       500 non-null object
Hqlocation     500 non-null object
Hqaddr         500 non-null object
Hqcity         500 non-null object
Hqstate        500 non-null object
Hqzip          500 non-null int64
Hqtel          500 non-null object
Ceo            500 non-null object
Ceo-title      500 non-null object
Address        500 non-null object
Ticker         500 non-null object
Fullname       500 non-null object
Revenues       500 non-null int64
Revchange      500 non-null float64
Profits        500 non-null float64
Prftchange     500 non-null object
Assets         500 non-null int64
Totshequity    500 non-null float64
dtypes: float64(3), int64(4), object(15)
memory usage: 89.8+ KB


## Pandas Data Selection - indexing


In [20]:
data = pd.Series(['a', 'b','c'], index = [1,3,5])

In [21]:
data

1    a
3    b
5    c
dtype: object

In [24]:
data[1:3]

3    b
5    c
dtype: object

### Selecting pandas data using “loc” (Selecting Columns From a DataFrame by Label)



<p><img alt="loc single column" src="https://s3.amazonaws.com/dq-content/291/loc_single.svg"></p>



In [25]:
f500.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 500 entries, 1 to 500
Data columns (total 22 columns):
Title          500 non-null object
Website        500 non-null object
Employees      500 non-null int64
Sector         500 non-null object
Industry       500 non-null object
Hqlocation     500 non-null object
Hqaddr         500 non-null object
Hqcity         500 non-null object
Hqstate        500 non-null object
Hqzip          500 non-null int64
Hqtel          500 non-null object
Ceo            500 non-null object
Ceo-title      500 non-null object
Address        500 non-null object
Ticker         500 non-null object
Fullname       500 non-null object
Revenues       500 non-null int64
Revchange      500 non-null float64
Profits        500 non-null float64
Prftchange     500 non-null object
Assets         500 non-null int64
Totshequity    500 non-null float64
dtypes: float64(3), int64(4), object(15)
memory usage: 89.8+ KB


In [26]:
industries = f500.loc[:, "Industry"]

In [28]:
industries.head()

Rank
1                       General Merchandisers
2    Insurance: Property and Casualty (Stock)
3                 Computers, Office Equipment
4                          Petroleum Refining
5                    Wholesalers: Health Care
Name: Industry, dtype: object


<p><img alt="loc list of columns" src="https://s3.amazonaws.com/dq-content/291/loc_list.svg"></p>

In [30]:
more_columns = f500.loc[:, ["Employees","Revenues", "Assets"]]

In [31]:
more_columns

Employees  Revenues   Assets
Rank                              
1       2300000    485873   198825
2        367700    223604   620854
3        116000    215639   321686
4         72700    205004   330314
5         68000    192487    56563
6        230000    184840   122810
7        204000    177526    94462
8        225000    166380   221690
9        268540    163786   403821
10       201000    151800   237951
11        18500    146850    33656
12       341400    135987    83402
13       295000    126661   365183
14       160900    125980   244180
15        37300    121546    34122
16       172000    118719    33163
17       300000    117351    72688
18       443000    115337    36505
19        55200    107567   260078
20         7000    107162  3287968
21       243355    105486  2490972
22        25600    100288    51745
23       406000     94595    42966
24       150540     94571    89997
25       269100     94176  1930115
26       208024     93662  2187702
27        72053     90272   167497
28       114000     85320   193694
29        53000     84863    65083
30       219000     82386  1792077
...         ...       ...      ...
471       23000      5505     9743
472         220      5488    11408
473        5930      5486     2219
474       32000      5457     6494
475        8000      5451    22615
476        4754      5449    20386
477        4050      5435    31104
478       10700      5423     9420
479       15700      5422   136453
480        1326      5414     2985
481       22600      5406     3010
482        7000      5400     6060
483        2178      5398    43831
484        7293      5389     8357
485       11870      5387    11770
486       11781      5379     4755
487        6600      5369     2164
488        3727      5354    22519
489       17000      5314     6135
490        7500      5277     2028
491        3626      5276    31377
492       19000      5259     7707
493       16400      5250     1778
494       20000      5236     2671
495       23150      5200     1608
496       31000      5197     2148
497        4200      5170     9737
498        8500      5169    48083
499        4431      5164    15167
500      110000      5145     2281

[500 rows x 3 columns]

In [33]:
financial_data = f500.loc[:, "Revenues" : "Profits"]
financial_data

Revenues  Revchange  Profits
Rank                              
1       485873        0.8  13643.0
2       223604        6.1  24074.0
3       215639       -7.7  45687.0
4       205004      -16.7   7840.0
5       192487        6.2   2258.0
6       184840       17.7   7017.0
7       177526       15.8   5317.0
8       166380        9.2   9427.0
9       163786       11.6  12976.0
10      151800        1.5   4596.0
11      146850        8.0   1427.9
12      135987       27.1   2371.0
13      126661       -9.8   8831.0
14      125980       -4.3  13127.0
15      121546       18.5   1427.0
16      118719        2.2   2350.0
17      117351       13.4   4173.0
18      115337        5.0   1975.0
19      107567      -18.0   -497.0
20      107162       -2.9  12313.0
21      105486        4.4  24733.0
22      100288       -1.4   3404.4
23       94595        6.9   7957.0
24       94571       -1.6   4895.0
25       94176        4.6  21938.0
26       93662        0.7  17906.0
27       90272       20.4  19478.0
28       85320       -8.8  16798.0
29       84863        7.2   2469.8
30       82386       -6.7  14912.0
...        ...        ...      ...
471       5505        4.8    930.0
472       5488        1.9    762.0
473       5486        2.1     84.7
474       5457       -4.3    318.0
475       5451       18.1    411.0
476       5449        6.0    590.7
477       5435        3.2   1838.9
478       5423       -1.1    253.2
479       5422       19.4    791.0
480       5414       -6.0   -153.7
481       5406        2.5    294.1
482       5400        2.5      7.0
483       5398        0.8    241.2
484       5389       -5.0    900.0
485       5387       -6.6   -383.5
486       5379       -3.8    312.0
487       5369        0.1     63.9
488       5354      -19.5  -1405.0
489       5314        0.2    337.5
490       5277       -4.4    120.8
491       5276       10.0    680.0
492       5259       -0.1    728.0
493       5250        3.1    343.4
494       5236       -2.3     15.0
495       5200        0.2    152.8
496       5197        5.8    378.2
497       5170       23.9    382.1
498       5169        4.0   -214.3
499       5164        4.0   -214.3
500       5145       -2.8     57.2

[500 rows x 3 columns]


<p><img alt="loc slice of columns" src="https://s3.amazonaws.com/dq-content/291/loc_slice.svg"></p>

In [36]:
industries = f500['Industry']
industries.head()

Rank
1                       General Merchandisers
2    Insurance: Property and Casualty (Stock)
3                 Computers, Office Equipment
4                          Petroleum Refining
5                    Wholesalers: Health Care
Name: Industry, dtype: object

In [40]:
industries = f500.Industry
industries.head()

Rank
1                       General Merchandisers
2    Insurance: Property and Casualty (Stock)
3                 Computers, Office Equipment
4                          Petroleum Refining
5                    Wholesalers: Health Care
Name: Industry, dtype: object


<p></p><center>
<table>
<thead>
<tr>
<th>Select by Label</th>
<th>Explicit Syntax</th>
<th>Common Shorthand</th>
<th>Other Shorthand</th>
</tr>
</thead>
<tbody>
<tr>
<td>Single column</td>
<td><code>df.loc[:,"col1"]</code></td>
<td><code>df["col1"]</code></td>
<td><code>df.col1</code></td>
</tr>
<tr>
<td>List of columns</td>
<td><code>df.loc[:,["col1", "col7"]]</code></td>
<td><code>df[["col1", "col7"]]</code></td>
<td></td>
</tr>
<tr>
<td>Slice of columns</td>
<td><code>df.loc[:,"col1":"col4"]</code></td>
<td></td>
<td></td>
</tr>
</tbody>
</table>
</center><p></p>




### Selecting Items from a Series by Label



<p><img alt="dataframe exploded" src="https://s3.amazonaws.com/dq-content/291/df_exploded.svg"></p>



<p><img alt="series vs dataframe: series" src="https://s3.amazonaws.com/dq-content/291/df_series_s.svg"></p>


<p><img alt="series vs dataframe: dataframe" src="https://s3.amazonaws.com/dq-content/291/df_series_df.svg"></p>



<center>
<table>
<thead>
<tr>
<th></th>
<th>Series</th>
<th>DataFrame</th>
</tr>
</thead>
<tbody>
<tr>
<th>Dimensions</th>
<td>One</td>
<td>Two</td>
</tr>
<tr>
<th>Has 'index' axis</th>
<td>Yes</td>
<td>Yes</td>
</tr>
<tr>
<th>Has 'columns' axis</th>
<td>No</td>
<td>Yes</td>
</tr>
<tr>
<th>Number of dtypes</th>
<td>One</td>
<td>Many (one per column)</td>
</tr>
</tbody>
</table>
</center>



In [44]:
s = pd.Series([4,7,-5,3], index = ['d','b','a','c'])

s
                                   

d    4
b    7
a   -5
c    3
dtype: int64

In [45]:
s['d']

4

In [46]:
s.keys()

Index(['d', 'b', 'a', 'c'], dtype='object')

In [55]:
ceos = f500['Ceo']
ceos.head(2)


SyntaxError: invalid syntax (<ipython-input-55-91ef04ff143f>, line 2)

### Selecting Rows From a DataFrame by Label



<p><img alt="anatomy of a dataframe" src="https://s3.amazonaws.com/dq-content/291/df_anatomy_static.svg"></p>



In [56]:
f500.index

Int64Index([  1,   2,   3,   4,   5,   6,   7,   8,   9,  10,
            ...
            491, 492, 493, 494, 495, 496, 497, 498, 499, 500],
           dtype='int64', name='Rank', length=500)

In [63]:
f500.set_index('Titles', inplace = True)


KeyError: 'Titles'

In [64]:
single_row = f500.loc['Walmart']
single_row

Website                               http://www.walmart.com
Employees                                            2300000
Sector                                             Retailing
Industry                               General Merchandisers
Hqlocation                                   Bentonville, AR
Hqaddr                                   702 S.W. Eighth St.
Hqcity                                           Bentonville
Hqstate                                                   AR
Hqzip                                                  72716
Hqtel                                           479-273-4000
Ceo                                      C. Douglas McMillon
Ceo-title      President, Chief Executive Officer & Director
Address           702 S.W. Eighth St., Bentonville, AR 72716
Ticker                                                   WMT
Fullname                               Wal-Mart Stores, Inc.
Revenues                                              485873
Revchange               

**pandas.DataFrame.set_index**

http://pandas.pydata.org/pandas-docs/stable/generated/pandas.DataFrame.set_index.html?highlight=set_index#pandas.DataFrame.set_index

`DataFrame.set_index(keys, drop=True, append=False, inplace=False, verify_integrity=False)`

Set the DataFrame index (row labels) using one or more existing columns. By default yields a new object.

<table>
<colgroup><col class="field-name">
<col class="field-body">
</colgroup><tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><dl class="first docutils">
<dt><strong>keys</strong> <span class="classifier-delimiter">:</span> <span class="classifier">column label or list of column labels / arrays</span></dt>
<dd></dd>
</dl>
<p><strong>drop</strong> : boolean, default True</p>
<blockquote>
<div><p>Delete columns to be used as the new index</p>
</div></blockquote>
<p><strong>append</strong> : boolean, default False</p>
<blockquote>
<div><p>Whether to append columns to existing index</p>
</div></blockquote>
<p><strong>inplace</strong> : boolean, default False</p>
<blockquote>
<div><p>Modify the DataFrame in place (do not create a new object)</p>
</div></blockquote>
<p><strong>verify_integrity</strong> : boolean, default False</p>
<blockquote>
<div><p>Check the new index for duplicates. Otherwise defer the check until
necessary. Setting to False will improve the performance of this
method</p>
</div></blockquote>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><dl class="first last docutils">
<dt><strong>dataframe</strong> <span class="classifier-delimiter">:</span> <span class="classifier">DataFrame</span></dt>
<dd></dd>
</dl>
</td>
</tr>
</tbody>
</table>

In [65]:
top_three = f500.loc
top_three

In [74]:
top_ = f500.loc[['Walmart':'Exxon Mobil'], ['Website' : 'Industry']]

SyntaxError: invalid syntax (<ipython-input-74-9405b8eedd33>, line 1)

In [76]:
top_ten = f500.iloc[:10,:]
top_ten.head(2)

Website  Employees      Sector  \
Title                                                                         
Walmart                       http://www.walmart.com    2300000   Retailing   
Berkshire Hathaway  http://www.berkshirehathaway.com     367700  Financials   

                                                    Industry       Hqlocation  \
Title                                                                           
Walmart                                General Merchandisers  Bentonville, AR   
Berkshire Hathaway  Insurance: Property and Casualty (Stock)        Omaha, NE   

                                 Hqaddr       Hqcity Hqstate  Hqzip  \
Title                                                                 
Walmart             702 S.W. Eighth St.  Bentonville      AR  72716   
Berkshire Hathaway      3555 Farnam St.        Omaha      NE  68131   

                           Hqtel  ...  \
Title                             ...   
Walmart             479-273-4000  ...   
Berkshire Hathaway  402-346-1400  ...   

                                                        Ceo-title  \
Title                                                               
Walmart             President, Chief Executive Officer & Director   
Berkshire Hathaway             Chairman & Chief Executive Officer   

                                                       Address Ticker  \
Title                                                                   
Walmart             702 S.W. Eighth St., Bentonville, AR 72716    WMT   
Berkshire Hathaway            3555 Farnam St., Omaha, NE 68131   BRKA   

                                   Fullname Revenues  Revchange  Profits  \
Title                                                                      
Walmart               Wal-Mart Stores, Inc.   485873        0.8  13643.0   
Berkshire Hathaway  Berkshire Hathaway Inc.   223604        6.1  24074.0   

                    Prftchange  Assets  Totshequity  
Title                                                
Walmart                   -7.2  198825      77798.0  
Berkshire Hathaway           0  620854     283001.0  

[2 rows x 21 columns]

In [77]:
top_ten.iloc[5:, 5:]

Hqaddr      Hqcity Hqstate  Hqzip  \
Title                                                                   
UnitedHealth Group        9900 Bren Rd. E.  Minnetonka      MN  55343   
CVS Health                       1 CVS Dr.  Woonsocket      RI   2895   
General Motors      300 Renaissance Center     Detroit      MI  48265   
AT&T                      208 S. Akard St.      Dallas      TX  75202   
Ford Motor                  1 American Rd.    Dearborn      MI  48126   

                           Hqtel                    Ceo  \
Title                                                     
UnitedHealth Group  952-936-1300     Stephen J. Hemsley   
CVS Health          401-765-1500         Larry J. Merlo   
General Motors      313-556-5000          Mary T. Barra   
AT&T                210-821-4105  Randall L. Stephenson   
Ford Motor          313-322-3000       James P. Hackett   

                                                        Ceo-title  \
Title                                                               
UnitedHealth Group             Chief Executive Officer & Director   
CVS Health          President, Chief Executive Officer & Director   
General Motors                 Chairman & Chief Executive Officer   
AT&T                Chairman, President & Chief Executive Officer   
Ford Motor          President, Chief Executive Officer & Director   

                                                      Address Ticker  \
Title                                                                  
UnitedHealth Group     9900 Bren Rd. E., Minnetonka, MN 55343    UNH   
CVS Health                    1 CVS Dr., Woonsocket, RI 02895    CVS   
General Motors      300 Renaissance Center, Detroit, MI 48265     GM   
AT&T                       208 S. Akard St., Dallas, TX 75202      T   
Ford Motor                 1 American Rd., Dearborn, MI 48126      F   

                                           Fullname  Revenues  Revchange  \
Title                                                                      
UnitedHealth Group  UnitedHealth Group Incorporated    184840       17.7   
CVS Health                   CVS Health Corporation    177526       15.8   
General Motors               General Motors Company    166380        9.2   
AT&T                                      AT&T Inc.    163786       11.6   
Ford Motor                       Ford Motor Company    151800        1.5   

                    Profits Prftchange  Assets  Totshequity  
Title                                                        
UnitedHealth Group   7017.0       20.7  122810      38274.0  
CVS Health           5317.0        1.5   94462      36830.0  
General Motors       9427.0       -2.7  221690      43836.0  
AT&T                12976.0       -2.8  403821     123135.0  
Ford Motor           4596.0      -37.7  237951      29170.0

In [79]:
top_ten.iloc[:, 0]

Title
Walmart                         http://www.walmart.com
Berkshire Hathaway    http://www.berkshirehathaway.com
Apple                             http://www.apple.com
Exxon Mobil                  http://www.exxonmobil.com
McKesson                       http://www.mckesson.com
UnitedHealth Group    http://www.unitedhealthgroup.com
CVS Health                    http://www.cvshealth.com
General Motors                       http://www.gm.com
AT&T                                http://www.att.com
Ford Motor                         http://www.ford.com
Name: Website, dtype: object

In [81]:
top_ten.iloc[9, :]

Website                                  http://www.ford.com
Employees                                             201000
Sector                                Motor Vehicles & Parts
Industry                            Motor Vehicles and Parts
Hqlocation                                      Dearborn, MI
Hqaddr                                        1 American Rd.
Hqcity                                              Dearborn
Hqstate                                                   MI
Hqzip                                                  48126
Hqtel                                           313-322-3000
Ceo                                         James P. Hackett
Ceo-title      President, Chief Executive Officer & Director
Address                   1 American Rd., Dearborn, MI 48126
Ticker                                                     F
Fullname                                  Ford Motor Company
Revenues                                              151800
Revchange               

### Selecting pandas data using “iloc”


## Series and Dataframe Describe Methods



In [82]:
f500.info()

<class 'pandas.core.frame.DataFrame'>
Index: 500 entries, Walmart to ABM Industries
Data columns (total 21 columns):
Website        500 non-null object
Employees      500 non-null int64
Sector         500 non-null object
Industry       500 non-null object
Hqlocation     500 non-null object
Hqaddr         500 non-null object
Hqcity         500 non-null object
Hqstate        500 non-null object
Hqzip          500 non-null int64
Hqtel          500 non-null object
Ceo            500 non-null object
Ceo-title      500 non-null object
Address        500 non-null object
Ticker         500 non-null object
Fullname       500 non-null object
Revenues       500 non-null int64
Revchange      500 non-null float64
Profits        500 non-null float64
Prftchange     500 non-null object
Assets         500 non-null int64
Totshequity    500 non-null float64
dtypes: float64(3), int64(4), object(14)
memory usage: 105.9+ KB


In [83]:
revs = f500['Revenues']
revs.describe()

count       500.000000
mean      24111.748000
std       38337.353337
min        5145.000000
25%        7245.000000
50%       11384.000000
75%       22605.250000
max      485873.000000
Name: Revenues, dtype: float64

In [84]:
f500['Assets'].describe()

count    5.000000e+02
mean     8.038934e+04
std      2.704257e+05
min      4.370000e+02
25%      8.436500e+03
50%      1.932450e+04
75%      4.812600e+04
max      3.287968e+06
Name: Assets, dtype: float64

In [85]:
f500['Hqstate'].describe()

count     500
unique     37
top        NY
freq       54
Name: Hqstate, dtype: object

In [86]:
f500.describe()

Employees         Hqzip       Revenues  Revchange       Profits  \
count  5.000000e+02    500.000000     500.000000  500.00000    500.000000   
mean   5.635013e+04  46791.714000   24111.748000    3.75800   1779.479800   
std    1.234520e+05  30160.384765   38337.353337   19.96748   3937.558721   
min    8.300000e+01   1104.000000    5145.000000  -57.50000  -6177.000000   
25%    1.190000e+04  19099.000000    7245.000000   -3.82500    235.725000   
50%    2.500000e+04  46244.500000   11384.000000    1.90000    683.600000   
75%    5.682525e+04  75045.000000   22605.250000    7.32500   1770.775000   
max    2.300000e+06  98188.000000  485873.000000  197.30000  45687.000000   

             Assets  Totshequity  
count  5.000000e+02     500.0000  
mean   8.038934e+04   13640.1470  
std    2.704257e+05   30523.1538  
min    4.370000e+02  -12688.0000  
25%    8.436500e+03    1997.5000  
50%    1.932450e+04    4981.0000  
75%    4.812600e+04   12467.7500  
max    3.287968e+06  283001.0000

In [88]:
f500.describe(include = ['O'])

Website      Sector                     Industry  \
count                     500         500                          500   
unique                    499          21                           73   
top     http://www.abbott.com  Financials  Utilities: Gas and Electric   
freq                        2          84                           22   

          Hqlocation                     Hqaddr    Hqcity Hqstate  \
count            500                        500       500     500   
unique           246                        491       239      37   
top     New York, NY  1211 Ave. of the Americas  New York      NY   
freq              45                          2        45      54   

               Hqtel              Ceo  \
count            500              500   
unique           499              496   
top     650-857-1501  Gregory J. Goff   
freq               2                2   

                                            Ceo-title  \
count                                             500   
unique                                             12   
top     President, Chief Executive Officer & Director   
freq                                              189   

                                         Address Ticker  \
count                                        500    500   
unique                                       494    468   
top     12300 Liberty Blvd., Englewood, CO 80112   NFLX   
freq                                           2      3   

                     Fullname Prftchange  
count                     500        500  
unique                    500        377  
top     Alleghany Corporation        5.2  
freq                        1          6

In [93]:
f500["Profits"].describe()

count      500.000000
mean      1779.479800
std       3937.558721
min      -6177.000000
25%        235.725000
50%        683.600000
75%       1770.775000
max      45687.000000
Name: Profits, dtype: float64

In [98]:
f500[['Revenues', 'Employees']].describe()

Revenues     Employees
count     500.000000  5.000000e+02
mean    24111.748000  5.635013e+04
std     38337.353337  1.234520e+05
min      5145.000000  8.300000e+01
25%      7245.000000  1.190000e+04
50%     11384.000000  2.500000e+04
75%     22605.250000  5.682525e+04
max    485873.000000  2.300000e+06

### More Data Exploration Methods



<p><img alt="dataframe axis parameters" src="https://s3.amazonaws.com/dq-content/291/axis_param.svg"></p>



In [99]:
df = pd.DataFrame({'A' : [1,1,2,1,2,3,3,3,3,3]})
df

A
0  1
1  1
2  2
3  1
4  2
5  3
6  3
7  3
8  3
9  3

In [101]:
df.sum(axis = 0)

A    22
dtype: int64

In [110]:
f500[['Revenues', "Profits"]].median(axis = 0)

Revenues    11384.0
Profits       683.6
dtype: float64

In [114]:
f500['Sector'].value_counts().head()

Financials     84
Energy         57
Retailing      47
Technology     43
Health Care    38
Name: Sector, dtype: int64

In [115]:
f500['Hqstate'].value_counts().head(10)

NY    54
CA    53
TX    50
IL    36
OH    25
VA    23
NJ    21
PA    21
MN    18
CT    18
Name: Hqstate, dtype: int64

In [122]:
f500.max(numeric_only = True)


Employees      2300000.0
Hqzip            98188.0
Revenues        485873.0
Revchange          197.3
Profits          45687.0
Assets         3287968.0
Totshequity     283001.0
dtype: float64

###  Assignment with pandas

In [124]:
top5_rank = f500[['Profits', 'Revenues']].head()
top5_rank

Profits  Revenues
Title                                
Walmart             13643.0    485873
Berkshire Hathaway  24074.0    223604
Apple               45687.0    215639
Exxon Mobil          7840.0    205004
McKesson             2258.0    192487

In [125]:
top5_rank['Revenues'] = 0
top5_rank

Profits  Revenues
Title                                
Walmart             13643.0         0
Berkshire Hathaway  24074.0         0
Apple               45687.0         0
Exxon Mobil          7840.0         0
McKesson             2258.0         0

In [129]:
top5_rank.iloc(['Apple']) = 99
top5_rank

SyntaxError: can't assign to function call (<ipython-input-129-bb9adf433bfd>, line 1)

In [130]:
f500['Revenues']

Title
Walmart                               485873
Berkshire Hathaway                    223604
Apple                                 215639
Exxon Mobil                           205004
McKesson                              192487
UnitedHealth Group                    184840
CVS Health                            177526
General Motors                        166380
AT&T                                  163786
Ford Motor                            151800
AmerisourceBergen                     146850
Amazon.com                            135987
General Electric                      126661
Verizon                               125980
Cardinal Health                       121546
Costco                                118719
Walgreens Boots Alliance              117351
Kroger                                115337
Chevron                               107567
Fannie Mae                            107162
J.P. Morgan Chase                     105486
Express Scripts Holding               100288
Home

In [132]:
f500['rev_bil'] = f500['Revenues'] / 1000
f500.head()

Website  Employees       Sector  \
Title                                                                          
Walmart                       http://www.walmart.com    2300000    Retailing   
Berkshire Hathaway  http://www.berkshirehathaway.com     367700   Financials   
Apple                           http://www.apple.com     116000   Technology   
Exxon Mobil                http://www.exxonmobil.com      72700       Energy   
McKesson                     http://www.mckesson.com      68000  Wholesalers   

                                                    Industry  \
Title                                                          
Walmart                                General Merchandisers   
Berkshire Hathaway  Insurance: Property and Casualty (Stock)   
Apple                            Computers, Office Equipment   
Exxon Mobil                               Petroleum Refining   
McKesson                            Wholesalers: Health Care   

                           Hqlocation                  Hqaddr         Hqcity  \
Title                                                                          
Walmart               Bentonville, AR     702 S.W. Eighth St.    Bentonville   
Berkshire Hathaway          Omaha, NE         3555 Farnam St.          Omaha   
Apple                   Cupertino, CA         1 Infinite Loop      Cupertino   
Exxon Mobil                Irving, TX  5959 Las Colinas Blvd.         Irving   
McKesson            San Francisco, CA              1 Post St.  San Francisco   

                   Hqstate  Hqzip         Hqtel  ...  \
Title                                            ...   
Walmart                 AR  72716  479-273-4000  ...   
Berkshire Hathaway      NE  68131  402-346-1400  ...   
Apple                   CA  95014  408-996-1010  ...   
Exxon Mobil             TX  75039  972-444-1000  ...   
McKesson                CA  94104  415-983-8300  ...   

                                                       Address Ticker  \
Title                                                                   
Walmart             702 S.W. Eighth St., Bentonville, AR 72716    WMT   
Berkshire Hathaway            3555 Farnam St., Omaha, NE 68131   BRKA   
Apple                     1 Infinite Loop, Cupertino, CA 95014   AAPL   
Exxon Mobil           5959 Las Colinas Blvd., Irving, TX 75039    XOM   
McKesson                   1 Post St., San Francisco, CA 94104    MCK   

                                   Fullname Revenues Revchange  Profits  \
Title                                                                     
Walmart               Wal-Mart Stores, Inc.   485873       0.8  13643.0   
Berkshire Hathaway  Berkshire Hathaway Inc.   223604       6.1  24074.0   
Apple                           Apple, Inc.   215639      -7.7  45687.0   
Exxon Mobil         Exxon Mobil Corporation   205004     -16.7   7840.0   
McKesson               McKesson Corporation   192487       6.2   2258.0   

                    Prftchange  Assets Totshequity  rev_bil  
Title                                                        
Walmart                   -7.2  198825     77798.0  485.873  
Berkshire Hathaway           0  620854    283001.0  223.604  
Apple                    -14.4  321686    128249.0  215.639  
Exxon Mobil              -51.5  330314    167325.0  205.004  
McKesson                    53   56563      8924.0  192.487  

[5 rows x 22 columns]

### Using Boolean Indexing with pandas Objects



<p><img alt="Boolean arrays in pandas" src="https://s3.amazonaws.com/dq-content/291/boolean_array_pandas.svg"></p>



<p><img alt="example dataframe" src="https://s3.amazonaws.com/dq-content/291/eg_df.svg"></p>



<p><img alt="boolean series" src="https://s3.amazonaws.com/dq-content/291/bool_series.svg"></p>



<p><img alt="boolean indexing dataframe" src="https://s3.amazonaws.com/dq-content/291/boolean_indexing_df.svg"></p>



<p><img alt="boolean indexing series" src="https://s3.amazonaws.com/dq-content/291/boolean_indexing_s.svg"></p>




In [ ]:
motor_boll = f500['Industry']

### Using Boolean Arrays to Assign Values



In [135]:
f500.loc['Industry'] == 'Motor vehicles & Parts'
f500

Website  \
Title                                                                       
Walmart                                            http://www.walmart.com   
Berkshire Hathaway                       http://www.berkshirehathaway.com   
Apple                                                http://www.apple.com   
Exxon Mobil                                     http://www.exxonmobil.com   
McKesson                                          http://www.mckesson.com   
UnitedHealth Group                       http://www.unitedhealthgroup.com   
CVS Health                                       http://www.cvshealth.com   
General Motors                                          http://www.gm.com   
AT&T                                                   http://www.att.com   
Ford Motor                                            http://www.ford.com   
AmerisourceBergen                        http://www.amerisourcebergen.com   
Amazon.com                                          http://www.amazon.com   
General Electric                                        http://www.ge.com   
Verizon                                            http://www.verizon.com   
Cardinal Health                             http://www.cardinalhealth.com   
Costco                                              http://www.costco.com   
Walgreens Boots Alliance            http://www.walgreensbootsalliance.com   
Kroger                                         http://www.thekrogerco.com   
Chevron                                            http://www.chevron.com   
Fannie Mae                                       http://www.fanniemae.com   
J.P. Morgan Chase                            http://www.jpmorganchase.com   
Express Scripts Holding                    http://www.express-scripts.com   
Home Depot                                       http://www.homedepot.com   
Boeing                                              http://www.boeing.com   
Wells Fargo                                     http://www.wellsfargo.com   
Bank of America Corp.                        http://www.bankofamerica.com   
Alphabet                                               http://www.abc.xyz   
Microsoft                                        http://www.microsoft.com   
Anthem                                           http://www.antheminc.com   
Citigroup                                        http://www.citigroup.com   
...                                                                   ...   
Host Hotels & Resorts                           http://www.hosthotels.com   
Insight Enterprises                                http://www.insight.com   
Mattel                                              http://www.mattel.com   
AmTrust Financial Services                    http://www.amtrustgroup.com   
Cincinnati Financial                                http://www.cinfin.com   
Simon Property Group                                 http://www.simon.com   
Western Union                                 http://www.westernunion.com   
KeyCorp                                                http://www.key.com   
Delek US Holdings                                  http://www.delekus.com   
Booz Allen Hamilton Holding                      http://www.boozallen.com   
Chemours                                          http://www.chemours.com   
Western & Southern Financial Group         http://www.westernsouthern.com   
Celanese                                          http://www.celanese.com   
Windstream Holdings                             http://www.windstream.com   
Seaboard                                      http://www.seaboardcorp.com   
Essendant                                        http://www.essendant.com   
Apache                                          http://www.apachecorp.com   
Airgas                                              http://www.airgas.com   
Kelly Services                               http://www.kellyservices.com   
Liberty Media                                 http://www.liber

### Deleting a Column from Your DataFrame


`DataFrame.drop(labels=None, axis=0, index=None, columns=None, level=None, inplace=False, errors='raise')`

Drop specified labels from rows or columns.

https://pandas.pydata.org/pandas-docs/stable/generated/pandas.DataFrame.drop.htmlhttps://pandas.pydata.org/pandas-docs/stable/generated/pandas.DataFrame.drop.html

Remove rows or columns by specifying label names and corresponding axis, or by specifying directly index or column names. When using a multi-index, labels on different levels can be removed by specifying the level.

In [137]:
df = pd.DataFrame(data=np.array([[1, 2, 3], [40, 50, 9], [7, 2, 9], [40, 50, 9], [23, 35, 37]]), 
                  index= [2.5, 12.6, 4.8, 4.8, 2.5], 
                  columns=['A', 'B', 'C'])


In [138]:
df

A   B   C
2.5    1   2   3
12.6  40  50   9
4.8    7   2   9
4.8   40  50   9
2.5   23  35  37

In [139]:
df.drop('A', axis = 'columns')

B   C
2.5    2   3
12.6  50   9
4.8    2   9
4.8   50   9
2.5   35  37

### Removing a Row from Your DataFrame




In [140]:
df

A   B   C
2.5    1   2   3
12.6  40  50   9
4.8    7   2   9
4.8   40  50   9
2.5   23  35  37

In [141]:
df_1 = df.drop_duplicates()
df_1

A   B   C
2.5    1   2   3
12.6  40  50   9
4.8    7   2   9
2.5   23  35  37

`DataFrame.drop_duplicates(subset=None, keep='first', inplace=False)`

Return DataFrame with duplicate rows removed, optionally only considering certain columns

<table class="docutils field-list" frame="void" rules="none">
<colgroup><col class="field-name">
<col class="field-body">
</colgroup><tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><p class="first"><strong>subset</strong> : column label or sequence of labels, optional</p>
<blockquote>
<div><p>Only consider certain columns for identifying duplicates, by
default use all of the columns</p>
</div></blockquote>
<p><strong>keep</strong> : {‘first’, ‘last’, False}, default ‘first’</p>
<blockquote>
<div><ul class="simple">
<li><code class="docutils literal notranslate"><span class="pre">first</span></code> : Drop duplicates except for the first occurrence.</li>
<li><code class="docutils literal notranslate"><span class="pre">last</span></code> : Drop duplicates except for the last occurrence.</li>
<li>False : Drop all duplicates.</li>
</ul>
</div></blockquote>
<p><strong>inplace</strong> : boolean, default False</p>
<blockquote>
<div><p>Whether to drop duplicates in place or to return a copy</p>
</div></blockquote>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><dl class="first last docutils">
<dt><strong>deduplicated</strong> <span class="classifier-delimiter">:</span> <span class="classifier">DataFrame</span></dt>
<dd></dd>
</dl>
</td>
</tr>
</tbody>
</table>

### Combining Datasets: Concat and Append


In [4]:
def make_df(cols, ind):
    """Quickly make a DataFrame"""
    data = {c: [str(c) + str(i) for i in ind]
            for c in cols}
    return pd.DataFrame(data, ind)



In [5]:
# example DataFrame
make_df('ABCD', range(3))

A   B   C   D
0  A0  B0  C0  D0
1  A1  B1  C1  D1
2  A2  B2  C2  D2

#### Concatenation of NumPy Arrays



In [6]:
x = [1, 2, 3]
y = [4, 5, 6]
z = [7, 8, 9]


#### Simple Concatenation with pd.concat



`pandas.concat(objs, axis=0, join='outer', join_axes=None, ignore_index=False, keys=None, levels=None, names=None, verify_integrity=False, sort=None, copy=True)`

[Dokumentacija](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.concat.html)


In [7]:
ser1 = pd.Series(['A', 'B', 'C'], index=[1, 2, 3])
ser2 = pd.Series(['D', 'E', 'F'], index=[4, 5, 6])


In [8]:
df1 = make_df('AB', [1, 2])
df2 = make_df('AB', [3, 4])



In [56]:
df3 = make_df('AB', [0, 1])
df4 = make_df('CD', [0, 1])

#### Duplicate indices



In [61]:
x = make_df('AB', [0, 1])
y = make_df('AB', [2, 3])

##### Catching the repeats as an error

##### Ignoring the index

##### Adding MultiIndex keys

[Več o Hierarchical Indexing](https://jakevdp.github.io/PythonDataScienceHandbook/03.05-hierarchical-indexing.html) 

#### Concatenation with joins

In [74]:
df5 = make_df('ABC', [1, 2])
df6 = make_df('BCD', [3, 4])

#### The append() method



### Aggregation and Grouping

#### Planets dataset



#### GroupBy: Split, Apply, Combine

##### Split, apply, combine

<img src="https://jakevdp.github.io/PythonDataScienceHandbook/figures/03.08-split-apply-combine.png" alt="">



In [9]:
df = pd.DataFrame({'key': ['A', 'B', 'C', 'A', 'B', 'C'],
                   'data': range(6)}, columns=['key', 'data'])


#### The GroupBy object

#### Column indexing

#### Dispatch methods



#### Aggregate, filter, transform, apply

In [10]:
rng = np.random.RandomState(0)
df = pd.DataFrame({'key': ['A', 'B', 'C', 'A', 'B', 'C'],
                   'data1': range(6),
                   'data2': rng.randint(0, 10, 6)},
                   columns = ['key', 'data1', 'data2'])


##### Aggregation


##### Filtering



##### Transformation

##### The apply() method

### Working with Time Series

#### Dates and Times in Python


##### Native Python dates and times



##### Typed arrays of times: NumPy's datetime64

[Dokumentacija](https://docs.scipy.org/doc/numpy/reference/arrays.datetime.html)

<div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<table>
<thead><tr>
<th>Code</th>
<th>Meaning</th>
<th>Time span (relative)</th>
<th>Time span (absolute)</th>
</tr>
</thead>
<tbody>
<tr>
<td><code>Y</code></td>
<td>Year</td>
<td>± 9.2e18 years</td>
<td>[9.2e18 BC, 9.2e18 AD]</td>
</tr>
<tr>
<td><code>M</code></td>
<td>Month</td>
<td>± 7.6e17 years</td>
<td>[7.6e17 BC, 7.6e17 AD]</td>
</tr>
<tr>
<td><code>W</code></td>
<td>Week</td>
<td>± 1.7e17 years</td>
<td>[1.7e17 BC, 1.7e17 AD]</td>
</tr>
<tr>
<td><code>D</code></td>
<td>Day</td>
<td>± 2.5e16 years</td>
<td>[2.5e16 BC, 2.5e16 AD]</td>
</tr>
<tr>
<td><code>h</code></td>
<td>Hour</td>
<td>± 1.0e15 years</td>
<td>[1.0e15 BC, 1.0e15 AD]</td>
</tr>
<tr>
<td><code>m</code></td>
<td>Minute</td>
<td>± 1.7e13 years</td>
<td>[1.7e13 BC, 1.7e13 AD]</td>
</tr>
<tr>
<td><code>s</code></td>
<td>Second</td>
<td>± 2.9e12 years</td>
<td>[ 2.9e9 BC, 2.9e9 AD]</td>
</tr>
<tr>
<td><code>ms</code></td>
<td>Millisecond</td>
<td>± 2.9e9 years</td>
<td>[ 2.9e6 BC, 2.9e6 AD]</td>
</tr>
<tr>
<td><code>us</code></td>
<td>Microsecond</td>
<td>± 2.9e6 years</td>
<td>[290301 BC, 294241 AD]</td>
</tr>
<tr>
<td><code>ns</code></td>
<td>Nanosecond</td>
<td>± 292 years</td>
<td>[ 1678 AD, 2262 AD]</td>
</tr>
<tr>
<td><code>ps</code></td>
<td>Picosecond</td>
<td>± 106 days</td>
<td>[ 1969 AD, 1970 AD]</td>
</tr>
<tr>
<td><code>fs</code></td>
<td>Femtosecond</td>
<td>± 2.6 hours</td>
<td>[ 1969 AD, 1970 AD]</td>
</tr>
<tr>
<td><code>as</code></td>
<td>Attosecond</td>
<td>± 9.2 seconds</td>
<td>[ 1969 AD, 1970 AD]</td>
</tr>
</tbody>
</table>

</div>
</div>

#### Dates and times in pandas: best of both worlds


#### Pandas Time Series: Indexing by Time

#### Pandas Time Series Data Structures


In [13]:
dates = pd.to_datetime([datetime(2015, 7, 3), '4th of July, 2015',
                       '2015-Jul-6', '07-07-2015', '20150708'])



#### Example

### Understanding SettingwithCopyWarning in pandas


#### What is SettingWithCopyWarning?



<img class="full-width" src="https://www.dataquest.io/wp-content/uploads/2019/01/view-vs-copy.png" alt="view-vs-copy">



<img class="full-width" src="https://www.dataquest.io/wp-content/uploads/2019/01/modifying.png" alt="modifying">



## Več o 

[Pandas on PyVideo](https://pyvideo.org/)